### PR TITLE
test.py: make sure topology/ and topology_custom/ passes with tablets on.

### PR DESCRIPTION
**Pull request by kostja**
Explicitly disable tablets in a few tests that rely on features not yet supported with tablets.

### CODE REVIEW

File: test/topology/test_mutation_schema_change.py [hunk 1]
```diff
@@ -20,6 +20,7 @@
 
 
 @pytest.mark.asyncio
+@pytest.mark.enable_tablets(False)
```
Comment (mykaul) [inline test/topology/test_mutation_schema_change.py:23]: Do we understand why they are not supported with Tablets? When will we enable them again? How can we track this?

File: test/topology/test_mv.py [hunk 2]
```diff
@@ -16,6 +16,8 @@
 
 from test.topology.util import new_test_keyspace, new_test_table, new_materialized_view
 
+ksdef = "WITH REPLICATION = { 'class': 'NetworkTopologyStrategy', 'replication_factor': 3 } AND TABLETS = {'enabled': false }"
```
Comment (kbr-scylla) [inline test/topology/test_mv.py:19]: I don't understand why tablets need to be disabled for this test. The commit message does not provide motivation. What is the connection between tombstone gc settings, and the necessity to disable tablets?

File: test/topology/test_mutation_schema_change.py [hunk 3]
```diff
@@ -81,6 +82,7 @@ async def test_mutation_schema_change(manager, random_tables):
...
```
Comment (kbr-scylla) [inline test/topology/test_mutation_schema_change.py:85]: What is the long-term plan -- one we support LWT in tablets, how would we enable this test in tablets mode? Preferably, we should eventually run the test in both modes.

I'd prefer if this became more explicit in the code. Even the reason why you disable tablets here is not explicitly stated, there is no comment or anything, you have to look into the git log to find out.

I propose to take the approach that @bitpathfinder took in `test_change_replication_factor_1_to_0`, where the test is parameterized in advance to run in both tablets and vnodes mode, but the tablets mode is currently disabled, with explicit reason provided (issue reference):
```python3
@pytest.mark.parametrize(
    "use_tablets",
    [
        pytest.param(False, id="vnodes"),
        pytest.param(True, id="tablets", marks=pytest.mark.xfail(reason="issue #20282")),
    ],
)
@pytest.mark.asyncio
async def test_change_replication_factor_1_to_0(request: pytest.FixtureRequest, manager: ManagerClient, use_tablets: bool) -> None:
```

**Review by kbr-scylla — CHANGES_REQUESTED** [review]

File: test/topology/test_mv.py (hunk 2 above)
Comment (kostja) [reply test/topology/test_mv.py:19]:
```The test file contains two test cases, which both test
materialized view tombstone gc settings. With tablets the default
is "repair" which is different from vnodes.
```
This is the commit message. The default in tablets and vnodes is different.  This alone is enough to break the logic of the test.
But the test is testing that the gc settings are not inhetrited. With tablets,the gc settings are forced. This is indistinguishable from inherit. 
 
I don't mean to be rude, but being busy, I don't think you bothered to expand and look at what the test is doing. This doesn't make sense for tablets.

Comment (kbr-scylla) [reply test/topology/test_mv.py:19]: > This alone is enough to break the logic of the test.

But the commit message doesn't explain why. It should provide a clear motivation.

> I don't mean to be rude, but being busy, I don't think you bothered to expand and look at what the test is doing. 

It's not the reviewer's job to reverse-engineer the motivation of the commit by reading the code. Instead, the motivation should be provided in the commit message, and then the reviewer confirms that the justification matches what the code is actually doing.

> the test is testing that the gc settings are not inhetrited. With tablets,the gc settings are forced. This is indistinguishable from inherit. (...) This doesn't make sense for tablets.

That's a great explanation. Why didn't you put it in the commit message?

File: test/topology/test_mutation_schema_change.py (hunk 3 above)
Comment (kostja) [reply test/topology/test_mutation_schema_change.py:85]: It is interesting that this comment is not consistent with the previous one.  Some tests are parameterized on tablets, and some just use a CREATE KEYSPACE statement in them, and use the default. In fact, most of the tests are not parameterized. 

We could try to make sure that every test is explicitly parameterized to use tablets or vnodes. Oh wait. We have it. it's --enable-tablets in test/pylib/scylla_cluster.py. 

So if we have a way to parameterize all tests, do we need to separately parameterize each test? And moreover, do we need to do it *before* it actually works? Shouldn't the one who will be fixing LWT, CDC, views then actually test the fix anyway, including by enabling some of the disabled tests? Do we need to replace a grep + sed with pytest.mark.parameterize? Is there a guarantee that the bug will get closed? Is there a guarantee that when the bug is closed the test is updated? 

Too much to consider for such a simple problem. 
So the the question arises, why do you bother to ask me to do it this way and not the other way if there is so much ambiguity which way is better? 

Is it a good time and place to come up wit ha systematic way to parameterize the test? After all, the purpose and scope of the pull request is to prepare the ground for enabling tablets across the board, as, I hope, 
we will do eventually for all tests in test.py. 
The pull request has no aim to solve the larger problem of sequal testing of vnode-based and tablet-based partitioning, which should be solved for all tests.

File: test/topology/test_mv.py (hunk 2 above)
Comment (kostja) [reply test/topology/test_mv.py:19]: Updated the comment.

File: test/topology/test_mutation_schema_change.py (hunk 1 above)
Comment (kostja) [reply test/topology/test_mutation_schema_change.py:23]: Added a comment.

**Comment by kbr-scylla** [comment]
@yaronkaikov seems that it's impossible to get a working CI run due to too many test runs scheduled. How can we proceed?

**Comment by yaronkaikov** [comment]
> @yaronkaikov seems that it's impossible to get a working CI run due to too many test runs scheduled. How can we proceed?

@kbr-scylla it seems that running all `topology` test suites (which we are doing based on this PR changes) takes 13 hours (repeating 10 times) !!!!, does it make sense? 

```
01:30:29  + bash -c 'set -o pipefail; ./tools/toolchain/dbuild -v `pwd`:/scylladir -v /jenkins/coredumps:/jenkins/coredumps -- ./test.py  --tmpdir=/scylladir/testlog/x86_64 --artifacts_dir_url https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12353/artifact --verbose --repeat 10 "topology" "topology" "topology/test_mutation_schema_change::*" "topology/test_mv::*" 2>&1 | tee tests_console_output_x86_64.txt'
01:30:33  Found 3290 tests.
.
.
.
.
14:30:49  [3290/3290] topology_random_failures  debug  [ PASS ] topology_random_failures.test_random_failures.10 69.54s
14:30:49  Stopping S3 mock server
14:30:49  ------------------------------------------------------------------------------
14:30:49  CPU utilization: 21.1%
```

And this is only the first iteration (we suppose to have 10)

**Comment by kbr-scylla** [comment]
@yaronkaikov it doesn't make sense. The whole next promotion pipeline takes <4h, and there we're running each test 3 times in each mode. Multiplying by 10/3 it should take less than 14h to run each test 10 times in each mode.

Are scylla-ci runs using different (slower) kinds of instances perhaps?

**Comment by yaronkaikov** [comment]
> @yaronkaikov it doesn't make sense. The whole next promotion pipeline takes <4h, and there we're running each test 3 times in each mode. Multiplying by 10/3 it should take less than 14h to run each test 10 times in each mode.
> 
> Are scylla-ci runs using different (slower) kinds of instances perhaps?

In https://github.com/scylladb/scylladb/pull/21070#issuecomment-2424026225 it also took 20 hours and ran on `spider8` which is a local (IL) strong machine, so something doesn't make sense at all.

**Comment by kbr-scylla** [comment]
@xtrey please take a look at this, why is it taking >20h to run all tests with --repeat 10? (When in next job it takes <4h with --repeat 3), maybe it's the way tests are scheduled?

**Comment by kbr-scylla** [comment]
anyway @yaronkaikov I propose to modify the scylla-ci jenkins pipeline so it's possible to provide a parameter to modify the number of repeats to use instead of the predefined 10 or 100

**Comment by yaronkaikov** [comment]
> anyway @yaronkaikov I propose to modify the scylla-ci jenkins pipeline so it's possible to provide a parameter to modify the number of repeats to use instead of the predefined 10 or 100

@kbr-scylla @kostja implementing it as part of https://github.com/scylladb/scylla-pkg/pull/4586

**Comment by denesb** [comment]
Conflicts, needs rebase.

File: test/pylib/random_tables.py [hunk 4]
```diff
@@ -266,8 +266,13 @@ class RandomTables():
     """A list of managed random tables"""
 
     def __init__(self, test_name: str, manager: ManagerClient, keyspace: str,
-                 replication_factor: int, dc_replication_factor: dict[str, int] = None):
+                 replication_factor: int,
+                 dc_replication_factor: dict[str, int] = None,
+                 enable_tablets: bool = None):
```
Comment (kbr-scylla) [inline test/pylib/random_tables.py:271]: nit: `None` is not a correct `bool`. If you want to use a type hint and support all of `None`, `True`, and `False`, the right type would be `None | bool` or equivalently `Optional[None]`.

File: test/topology_custom/test_group0_schema_versioning.py [hunk 5]
```diff
@@ -124,7 +124,8 @@ async def test_schema_versioning_with_recovery(manager: ManagerClient):
     Verify that schema versions are in sync after each schema change.
     """
     cfg = {'enable_user_defined_functions': False,
-           'force_gossip_topology_changes': True}
+           'force_gossip_topology_changes': True,
+           'enable_tablets': False}
```
Comment (kbr-scylla) [inline test/topology_custom/test_group0_schema_versioning.py:128]: This should be implicit from `force_gossip_topology_changes`. What happens when you skip it? Is there an exception?

File: test/topology_custom/test_view_build_status.py [hunk 6]
```diff
@@ -192,7 +192,7 @@ async def test_view_build_status_snapshot(manager: ManagerClient):
 @pytest.mark.asyncio
 async def test_view_build_status_migration_to_v2(request, manager: ManagerClient):
     # First, force the first node to start in legacy mode
-    cfg = {'force_gossip_topology_changes': True}
+    cfg = {'force_gossip_topology_changes': True, 'enable_tablets': False}
```
Comment (kbr-scylla) [inline test/topology_custom/test_view_build_status.py:195]: again it's weird that this needs to be specified at all, it should be implicit from `force_gossip_topology_changes`. Maybe we have a bug.

File: test/topology_custom/test_mv_topology_change.py [hunk 7]
```diff
@@ -29,7 +29,9 @@
 @pytest.mark.asyncio
 @skip_mode('release', 'error injections are not supported in release mode')
 async def test_mv_topology_change(manager: ManagerClient):
-    cfg = {'force_gossip_topology_changes': True, 'error_injections_at_startup': ['delay_before_get_view_natural_endpoint']}
+    cfg = {'force_gossip_topology_changes': True,
+           'enable_tablets': False,
```
Comment (kbr-scylla) [inline test/topology_custom/test_mv_topology_change.py:33]: same here

File: test/topology_custom/test_raft_recovery_basic.py [hunk 8]
```diff
@@ -20,7 +20,8 @@
 @log_run_time
 async def test_raft_recovery_basic(request, manager: ManagerClient):
     cfg = {'enable_user_defined_functions': False,
-           'force_gossip_topology_changes': True}
+           'force_gossip_topology_changes': True,
+           'enable_tablets': False}
```
Comment (kbr-scylla) [inline test/topology_custom/test_raft_recovery_basic.py:24]: and so on

**Review by kbr-scylla — APPROVED** [review]
A nit and a question, but nothing blocking the PR

File: test/pylib/random_tables.py (hunk 4 above)
Comment (kostja) [reply test/pylib/random_tables.py:271]: Done.

File: test/topology_custom/test_group0_schema_versioning.py (hunk 5 above)
Comment (kostja) [reply test/topology_custom/test_group0_schema_versioning.py:128]: At the time this pull request was submitted htere was no check., so one would get runtime errors. It was addressed separately by @bhalevy in his patch for https://github.com/scylladb/scylladb/issues/21075

File: test/topology_custom/test_view_build_status.py (hunk 6 above)
Comment (kostja) [reply test/topology_custom/test_view_build_status.py:195]: https://github.com/scylladb/scylladb/issues/21075

File: test/topology_custom/test_mv_topology_change.py (hunk 7 above)
Comment (kostja) [reply test/topology_custom/test_mv_topology_change.py:33]: https://github.com/scylladb/scylladb/issues/21075

**Comment by kostja** [comment]
Updated the nit & rebased. CC @denesb

**Pull request closed without merging**